src/client/input/sdl.c

```diff
@@ -1360,6 +1382,371 @@ In_FlushQueue(void)
+		last_haptic_effect[i].effect_z = 0;
+
+		IN_Haptic_Effect_Shutdown(&last_haptic_effect[i].effect_id);
+	}
+}
+
+static int
+IN_Haptic_GetEffectId(int effect_volume, int effect_duration,
+				int effect_delay, int effect_attack, int effect_fade,
+				int effect_x, int effect_y, int effect_z)
+{
+	int i, haptic_volume;
+
+	// check effects for reuse
+	for (i=0; i < last_haptic_effect_size; i++)
+	{
+		if (
+			last_haptic_effect[i].effect_volume != effect_volume ||
+			last_haptic_effect[i].effect_duration != effect_duration ||
+			last_haptic_effect[i].effect_delay != effect_delay ||
+			last_haptic_effect[i].effect_attack != effect_attack ||
+			last_haptic_effect[i].effect_fade != effect_fade ||
+			last_haptic_effect[i].effect_x != effect_x ||
+			last_haptic_effect[i].effect_y != effect_y ||
+			last_haptic_effect[i].effect_z != effect_z)
+		{
+			continue;
+		}
+
+		return last_haptic_effect[i].effect_id;
+	}
+
+	/* create new effect */
+	haptic_volume = joy_haptic_magnitude->value * effect_volume; // 32767 max strength;
+
+	/*
+	Com_Printf("%d: volume %d: %d ms %d:%d:%d ms speed: %.2f\n",
+		last_haptic_effect_pos,  effect_volume, effect_duration,
+		effect_delay, effect_attack, effect_fade,
+		(float)effect_volume / effect_fade);
+	*/
+
+	// FIFO for effects
+	last_haptic_effect_pos = (last_haptic_effect_pos + 1) % last_haptic_effect_size;
+	IN_Haptic_Effect_Shutdown(&last_haptic_effect[last_haptic_effect_pos].effect_id);
+	last_haptic_effect[last_haptic_effect_pos].effect_volume = effect_volume;
+	last_haptic_effect[last_haptic_effect_pos].effect_duration = effect_duration;
+	last_haptic_effect[last_haptic_effect_pos].effect_delay = effect_delay;
+	last_haptic_effect[last_haptic_effect_pos].effect_attack = effect_attack;
+	last_haptic_effect[last_haptic_effect_pos].effect_fade = effect_fade;
+	last_haptic_effect[last_haptic_effect_pos].effect_x = effect_x;
+	last_haptic_effect[last_haptic_effect_pos].effect_y = effect_y;
+	last_haptic_effect[last_haptic_effect_pos].effect_z = effect_z;
+	last_haptic_effect[last_haptic_effect_pos].effect_id = IN_Haptic_Effect_Init(
+		effect_x, effect_y, effect_z,
+		effect_duration, haptic_volume,
+		effect_delay, effect_attack, effect_fade);
+
+	return last_haptic_effect[last_haptic_effect_pos].effect_id;
+}
+
+// Keep it same with rumble rules, look for descriptions to rumble
+// filtering in Controller_Rumble
+static char *default_haptic_filter = (
+	// skipped files should be before wider rule
+	"!weapons/*grenlb "     // bouncing grenades don't have feedback
+	"!weapons/*hgrenb "     // bouncing grenades don't have feedback
+	"!weapons/*open "       // rogue's items don't have feedback
+	"!weapons/*warn "       // rogue's items don't have feedback
+	// any weapons that are not in previous list
+	"weapons/ "
+	// player{,s} effects
+	"player/*land "         // fall without injury
+	"player/*burn "
+	"player/*pain100 "
+	"player/*pain75 "
+	"player/*pain50 "
+	"player/*pain25 "
+	"player/*pain25 "
```

Comment:
The last two are repeated. But you can leave this even more succint: by replacing all of these with `"player/*pain "`. You don't need to define every type of pain individually, since you are not specifying any difference between them.
A few lines below, in `players`, the same happens.

Reply:
Thank you, fixed